Unhandled panic in event collector

```
E0704 22:02:53.362781 1 runtime.go:79] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x3884720), concrete:(*runtime._type)(nil), asserted:(*runtime._type)(0x3f77de0), missingMethod:""} (interface conversion: interface {} is nil, not *v1.Event)
goroutine 230 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3995740?, 0xc002305650})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002305620?})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:49 +0x75
panic({0x3995740, 0xc002305650})
/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/rancher/opni/pkg/events.(*EventCollector).processItem(0xc000e0e7e0, {0x3af4140?, 0xc002305620?})
/src/pkg/events/events.go:177 +0x118
github.com/rancher/opni/pkg/events.(*EventCollector).processNextItem(0xc000e0e7e0)
/src/pkg/events/events.go:153 +0xf4
github.com/rancher/opni/pkg/events.(*EventCollector).runWorker(0xc000f6dd08?)
/src/pkg/events/events.go:131 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xb2b09734d4?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000f6ddd8?, {0x4e5c180, 0xc000c87e30}, 0x1, 0x0)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0xd0?, 0x73eac00?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:90
github.com/rancher/opni/pkg/events.(*EventCollector).Run(0xc000e0e7e0, 0x0)
/src/pkg/events/events.go:124 +0x425
github.com/rancher/opni/pkg/opni/commands.runEventsCollector({0x4e8a1d0, 0xc000126000})
/src/pkg/opni/commands/agent.go:311 +0x5b
github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2.3({0x4e8a1d0?, 0xc000126000?})
/src/pkg/opni/commands/agent.go:89 +0x65
created by github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2
/src/pkg/opni/commands/agent.go:87 +0x40a
E0704 22:02:53.362872 1 runtime.go:79] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x3884720), concrete:(*runtime._type)(nil), asserted:(*runtime._type)(0x3f77de0), missingMethod:""} (interface conversion: interface {} is nil, not *v1.Event)
goroutine 230 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3995740?, 0xc002305650})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0012bd8e0?})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:49 +0x75
panic({0x3995740, 0xc002305650})
/usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002305620?})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:56 +0xd8
panic({0x3995740, 0xc002305650})
/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/rancher/opni/pkg/events.(*EventCollector).processItem(0xc000e0e7e0, {0x3af4140?, 0xc002305620?})
/src/pkg/events/events.go:177 +0x118
github.com/rancher/opni/pkg/events.(*EventCollector).processNextItem(0xc000e0e7e0)
/src/pkg/events/events.go:153 +0xf4
github.com/rancher/opni/pkg/events.(*EventCollector).runWorker(0xc000f6dd08?)
/src/pkg/events/events.go:131 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xb2b09734d4?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000f6ddd8?, {0x4e5c180, 0xc000c87e30}, 0x1, 0x0)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0xd0?, 0x73eac00?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:90
github.com/rancher/opni/pkg/events.(*EventCollector).Run(0xc000e0e7e0, 0x0)
/src/pkg/events/events.go:124 +0x425
github.com/rancher/opni/pkg/opni/commands.runEventsCollector({0x4e8a1d0, 0xc000126000})
/src/pkg/opni/commands/agent.go:311 +0x5b
github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2.3({0x4e8a1d0?, 0xc000126000?})
/src/pkg/opni/commands/agent.go:89 +0x65
created by github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2
/src/pkg/opni/commands/agent.go:87 +0x40a
panic: interface conversion: interface {} is nil, not *v1.Event [recovered]
panic: interface conversion: interface {} is nil, not *v1.Event [recovered]
panic: interface conversion: interface {} is nil, not *v1.Event
goroutine 230 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0012bd8e0?})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:56 +0xd8
panic({0x3995740, 0xc002305650})
/usr/local/go/src/runtime/panic.go:838 +0x207
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002305620?})
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/runtime/runtime.go:56 +0xd8
panic({0x3995740, 0xc002305650})
/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/rancher/opni/pkg/events.(*EventCollector).processItem(0xc000e0e7e0, {0x3af4140?, 0xc002305620?})
/src/pkg/events/events.go:177 +0x118
github.com/rancher/opni/pkg/events.(*EventCollector).processNextItem(0xc000e0e7e0)
/src/pkg/events/events.go:153 +0xf4
github.com/rancher/opni/pkg/events.(*EventCollector).runWorker(0xc000f6dd08?)
/src/pkg/events/events.go:131 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xb2b09734d4?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:155 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000f6ddd8?, {0x4e5c180, 0xc000c87e30}, 0x1, 0x0)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0xd0?, 0x73eac00?)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
/root/.cache/go-mod/k8s.io/apimachinery@v0.24.1/pkg/util/wait/wait.go:90
github.com/rancher/opni/pkg/events.(*EventCollector).Run(0xc000e0e7e0, 0x0)
/src/pkg/events/events.go:124 +0x425
github.com/rancher/opni/pkg/opni/commands.runEventsCollector({0x4e8a1d0, 0xc000126000})
/src/pkg/opni/commands/agent.go:311 +0x5b
github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2.3({0x4e8a1d0?, 0xc000126000?})
/src/pkg/opni/commands/agent.go:89 +0x65
created by github.com/rancher/opni/pkg/opni/commands.BuildAgentCmd.func2
/src/pkg/opni/commands/agent.go:87 +0x40a
```

--- a/apis/v1beta2/logadapter_webhook.go
+++ b/apis/v1beta2/logadapter_webhook.go
@@ -18,9 +18,9 @@ package v1beta2
 
 import (
 	"fmt"
-	"reflect"
 
 	loggingv1beta1 "github.com/banzaicloud/logging-operator/pkg/sdk/logging/api/v1beta1"
+	"github.com/rancher/opni/pkg/util"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/intstr"
@@ -171,20 +171,16 @@ func validateProviderSettings(r *LogAdapter) error {
 
 	// Ensure that the provider matches the given settings
 	for k, v := range fields {
-		if (v == nil || isInterfaceNil(v)) && r.Spec.Provider == k {
+		if (v == nil || util.IsInterfaceNil(v)) && r.Spec.Provider == k {
 			return field.Required(field.NewPath("spec", "provider"),
 				"Provider settings field was not defaulted correctly (is the webhook running?)")
-		} else if !(v == nil || isInterfaceNil(v)) && r.Spec.Provider != k {
+		} else if !(v == nil || util.IsInterfaceNil(v)) && r.Spec.Provider != k {
 			return field.Forbidden(field.NewPath("spec", "provider"),
 				fmt.Sprintf("Provider is set to %s, but field %s is set, value is %v", r.Spec.Provider, field.NewPath("spec", string(k)), v))
 		}
 	}
 
 	return nil
-}
-
-func isInterfaceNil(i interface{}) bool {
-	return reflect.ValueOf(i).Kind() == reflect.Ptr && reflect.ValueOf(i).IsNil()
 }
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type

--- a/pkg/events/events.go
+++ b/pkg/events/events.go
@@ -8,6 +8,7 @@ import (
 
 	"github.com/opensearch-project/opensearch-go/opensearchutil"
 	"github.com/rancher/opni/pkg/logger"
+	"github.com/rancher/opni/pkg/util"
 	"go.uber.org/zap"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -172,6 +173,11 @@ func (c *EventCollector) processItem(obj interface{}) error {
 	event, _, err := c.informer.Informer().GetIndexer().GetByKey(eventObj.key)
 	if err != nil {
 		return err
+	}
+
+	if event == nil || util.IsInterfaceNil(event) {
+		c.logger.Info("nil event, skipping")
+		return nil
 	}
 
 	return c.shipEvent(event.(*corev1.Event), eventObj.time)

--- a/pkg/events/events.go
+++ b/pkg/events/events.go
@@ -27,6 +27,7 @@ type EventOutput struct {
 	*corev1.Event
 	ClusterID string    `json:"cluster_id,omitempty"`
 	LogType   string    `json:"log_type,omitempty"`
+	PodName   string    `json:"pod_name,omitempty"`
 	Time      time.Time `json:"time,omitempty"`
 }
 
@@ -194,7 +195,13 @@ func (c *EventCollector) shipEvent(event *corev1.Event, timestamp time.Time) err
 			Event:     event,
 			ClusterID: string(systemNamespace.GetUID()),
 			LogType:   "event",
-			Time:      timestamp,
+			PodName: func() string {
+				if event.InvolvedObject.Kind == "Pod" {
+					return event.InvolvedObject.Name
+				}
+				return ""
+			}(),
+			Time: timestamp,
 		},
 	}
 

--- a/pkg/util/generic.go
+++ b/pkg/util/generic.go
@@ -69,3 +69,7 @@ func Pointer[T any](t T) *T {
 func ProtoClone[T proto.Message](msg T) T {
 	return proto.Clone(msg).(T)
 }
+
+func IsInterfaceNil(i interface{}) bool {
+	return reflect.ValueOf(i).Kind() == reflect.Ptr && reflect.ValueOf(i).IsNil()
+}

